add a keybindings popup

TeFiLeDo opened the pull request:
Resolves #47.

Currently the content of the popup is just a paragraph. It might be worthwhile to change this, so that it uses the `Table` widget. However, to do this I'd need a bit more time, as I currently don't really understand how that would work.

I'm also thinking about adding a build script to extract the keybindings table from the `README.md`, so the information doesn't get out of sync.

I added the popup shortcuts at the top to give them the best chance of being visible in a small terminal.

konradsz commented:
I am impressed, thank you 🥇 

I have following ideas in mind (sorted by "priority"):
- it would be cool to have any key pressed closing the help popup (maybe with title on a bottom saying "Press any key to close...")
- you could calculate popup area dynamically by checking the length of the longest line in a text and also the number of lines and use it directly (with respect to window size) rather than percentages. This way the text in a popup would seem centered.
- add "header" line to the popup with bolded and centered "igrep v1.2.0" (can be built using `CARGO_PKG_NAME` and `CARGO_PKG_VERSION` env variables) - this needs some experiments to see if it looks good or not

Build script for extracting the text from README sounds great, but you can always do it in a separate PR if you fancy for a challenge :)

Let me know if you have time and interest in applying my suggestions, I can always merge it as is (cause it does its job!) and do it in the future.

konradsz commented on src/ui/input_handler.rs:
```diff
@@ -92,14 +102,73 @@ impl InputHandler {
+                code: KeyCode::Char('c'),
+                modifiers: KeyModifiers::CONTROL,
+            } => {
+                self.input_mode = InputMode::Normal;
```
using Esc or Ctrl+c when popup is visible closes application and I think your intent was to close popup only, am I right?

TeFiLeDo replied:
That was the intention, yes. IIRC it seemed to work, when I tested it. Same goes for the new version (where all other keys close the popup).

If there is some edge case, I think it would also affect the search popup, since I largely copied what it does here.

konradsz replied:
it must have been my terminal or something, because the issue disappear 🤷‍♂️ I am going to observe it

TeFiLeDo commented:
1. Added the "any key to close" behavior. Should that line specify that scrolling keys aren't affected?
2. The popup now uses as much space as it needs. It still is capped at 80% on small terminals, so that it is always identifiable as a popup.
3. I added the the `igrep v1.2.0` to the title of the popup. I don't think it looks good as part of the popup content.
4. Extraction via build script is now implemented.

konradsz commented on src/ui/keymap_popup.rs:
```diff
@@ -0,0 +1,106 @@
+                    .borders(Borders::ALL)
+                    .border_style(theme.search_popup_border())
+                    .title(concat!(
+                        "Keybindings - ",
```
I like the idea to have it in a title, but I let's skip "Keybindings" part, let it be simply " igrep v.1.2.0 " (with surrounding spaces :))

TeFiLeDo replied:
Done

konradsz commented on src/ui/keymap_popup.rs:
```diff
@@ -0,0 +1,106 @@
+                        env!("CARGO_PKG_VERSION"),
+                    ))
+                    .title_alignment(Alignment::Center)
+                    .padding(Padding::horizontal(1)),
```
```suggestion
                    .padding(Padding::uniform(1)),
```
IMO it looks better then. You need to change the calculation of vertical size then.

TeFiLeDo replied:
Done

konradsz commented:
There is one bug in your scrolling implementation. If you scroll down to the very bottom and then press "j" or "down" a number of times, then you need to press "k" or "up" the same number of times before it starts scrolling back :)

The rest looks great!

TeFiLeDo commented:
Scrolling should be fixed now
